flash/data/data_source.py

```diff
@@ -0,0 +1,264 @@
+                    data = load_data(data, mock_dataset)
+                else:
+                    data = load_data(data)
+
+            if has_len(data):
+                dataset = AutoDataset(data, self, running_stage)
+            else:
+                dataset = IterableAutoDataset(data, self, running_stage)
+            dataset.__dict__.update(mock_dataset.metadata)
+            return dataset
+
+
+class DefaultDataSources(LightningEnum):
+
+    PATHS = "paths"
+    NUMPY = "numpy"
+    TENSOR = "tensor"
+    CSV = "csv"
+    JSON = "json"
+
+    # TODO: Create a FlashEnum class???
+    def __hash__(self) -> int:
+        return hash(self.value)
+
+
+class DefaultDataKeys(LightningEnum):
+
+    INPUT = "input"
+    TARGET = "target"
+
+    # TODO: Create a FlashEnum class???
+    def __hash__(self) -> int:
+        return hash(self.value)
+
+
+SEQUENCE_DATA_TYPE = TypeVar("SEQUENCE_DATA_TYPE")
+
+
+class SequenceDataSource(
+    Generic[SEQUENCE_DATA_TYPE],
+    DataSource[Tuple[Sequence[SEQUENCE_DATA_TYPE], Optional[Sequence[Any]]]],
+):
+
+    def __init__(self, labels: Optional[Sequence[str]] = None):
+        super().__init__()
+
+        self.labels = labels
+
+        if self.labels is not None:
+            self.set_state(LabelsState(self.labels))
+
+    def load_data(
+        self,
+        data: Tuple[Sequence[SEQUENCE_DATA_TYPE], Optional[Sequence[Any]]],
+        dataset: Optional[Any] = None,
+    ) -> Sequence[Mapping[str, Any]]:
+        # TODO: Bring back the code to work out how many classes there are
+        inputs, targets = data
+        if targets is None:
+            return self.predict_load_data(data)
+        return [{
+            DefaultDataKeys.INPUT: input,
+            DefaultDataKeys.TARGET: target
+        } for input, target in zip(inputs, targets)]
+
+    def predict_load_data(self, data: Sequence[SEQUENCE_DATA_TYPE]) -> Sequence[Mapping[str, Any]]:
+        return [{DefaultDataKeys.INPUT: input} for input in data]
+
+
+class PathsDataSource(SequenceDataSource):  # TODO: Sort out the typing here
+
+    def __init__(self, extensions: Optional[Tuple[str, ...]] = None):
+        super().__init__()
+
+        self.extensions = extensions
+
+    @staticmethod
+    def find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
+        """
+        Finds the class folders in a dataset. Ensures that no class is a subdirectory of another.
+
+        Args:
+            dir: Root directory path.
+
+        Returns:
+            tuple: (classes, class_to_idx) where classes are relative to (dir), and class_to_idx is a dictionary.
+        """
+        classes = [d.name for d in os.scandir(dir) if d.is_dir()]
+        classes.sort()
+        class_to_idx = {cls_name: i for i, cls_name in enumerate(classes)}
+        return classes, class_to_idx
+
+    @staticmethod
+    def isdir(data: Union[str, Tuple[List[str], List[Any]]]) -> bool:
+        try:
+            return os.path.isdir(data)
+        except TypeError:
```

Comment:
in which case this will fail ?

Reply:
I've added a comment explaining, but is just duck typing in case it is a list of paths